### PR TITLE
Add ThrowsCasing for weapon casing throwing

### PR DESCRIPTION
**Pull request by dnqbob**
@EoralMilk version of #20395, Supersedes #20395

casing-showcase

### CODE REVIEW

**Comment by dnqbob** [comment]
The benefit from `BarrelIndex`: modeling `Armament`.

With `BarrelIndex` we can also separate the `MuzzleSquence` and `MuzzlePalette` from `Armament` into `WithMuzzleOverlay`. We also has a similar muzzle trait in TA using `BarrelIndex`.

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 1]
```diff
@@ -40,12 +41,16 @@ public class ArmamentInfo : PausableConditionalTraitInfo, Requires<AttackBaseInf
...
+			"If weapon Burst = 1, it cycles through all listed offsets, otherwise the offset corresponding to current burst is used.",
+			"Every position here will generate a barrel in this armament.")]
 		public readonly WVec[] LocalOffset = Array.Empty<WVec>();
 
-		[Desc("Muzzle yaw relative to turret or body.")]
+		[Desc("Muzzle yaw relative to turret or body on generated barrel.")]
 		public readonly WAngle[] LocalYaw = Array.Empty<WAngle>();
 
+		[Desc("Allow player to set the barrel index manually if needed.")]
+		public readonly int[] BarrelsIndex = Array.Empty<int>();
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/Armament.cs:52]: Why is this needed exactly? I don't see a point in in setting it manually

**Comment by PunkPun** [comment]
First and second commits are fine, but about the 3rd. If we aren't implementing `ThrowsCasing` trait anywhere in the base mods then I don't really think we can accept it into the code base. Besides, with the first and second commits merged, SDK mods could just implement that trait themselves without needing to modify the engine. So that's a win on its own.

If anyone knows some missing visual feature in the base mods that could solved via this trait then I'd be happy to merge it

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 1 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Armament.cs:52]: This will provide a convenient YAML writing for something like:

1. Fire at the same time
```
Armament@1:
    Weapon: cannon1
    Name: primary
    BarrelsIndex: 0, 1
Armament@2:
    Weapon: cannon1
    Name: primary
    BarrelsIndex: 1, 0
ThrowsCasing: 
    Armaments: primary
    Weapon: casing1
    CasingOffset: [offset0], [offset1]
```
Or you have to write 2 `ThrowsCasing` trait.

2. More than one fire effect for one Armament
```
Armament@1:
    Weapon: MRLSlauncher1
    Name: primary
    BarrelsIndex: 1, 0, 1, 0
ThrowsCasing@1: 
    Armaments: primary
    Weapon: casing1
    CasingOffset: [offset0], [offset1]
ThrowsCasing@2: 
    Armaments: primary
    Weapon: casing2
    CasingOffset: [offset0], [offset1]
```
Or you have to write  [offset0], [offset1],  [offset3], [offset4] for each `ThrowsCasing`

**Comment by dnqbob** [comment]
> First and second commits are fine, but about the 3rd. If we aren't implementing `ThrowsCasing` trait anywhere in the base mods then I don't really think we can accept it into the code base. Besides, with the first and second commits merged, SDK mods could just implement that trait themselves without needing to modify the engine. So that's a win on its own.
> 

this one is OK, I can just drop the `ThrowCasing` later, or you can just merge 1 and 2, consider 3 is a testcase.

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Armament.cs:52]: Oh I see, this is for firing multiple barrels in the same frame. But wouldn't Burst: 2 BurstDelay 0 achieve the same result and be much clearer to modders?

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Armament.cs:52]: Regretably, no. "BurstDelay=0" will still result in burst fire at the next frame. I filled up #20362 to add this function but I have no idea on making it work normal with `AmmoPool`

Comment (PunkPun) [reply OpenRA.Mods.Common/Traits/Armament.cs:52]: Then we should make it work with 0, or think of another solution. The `BarrelsIndex:` implementation feels more like a hack and is fairly confusing. You have to really study the code to understand how it works

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Armament.cs:52]: Yes and this one is tricky, but it brings some prefs and code simplify.

**Comment by AspectInteractive2** [comment]
First commit I think is fine to go in, as it makes the muzzle code slightly more extensible.

For the second commit the `BarrelIndex` is confusing in its current state and doesn't benefit the current code base, since the added complexity is only beneficial for the `ThrowsCasing` commit which does not fit within any of the existing mods. I also agree with Pun that if the motivation here is to allow simultaneous firing of barrels, this should be fixed in `Armament` as you attempted in #20362 rather than adding an extra array to deal with the problem (this seems hacky, as Pun mentioned).

**Comment by dnqbob** [comment]
#20601 as reopened https://github.com/OpenRA/OpenRA/pull/20362, ammopool problem fixed

**Comment by Mailaender** [comment]
We implemented a similar effect without the need for any changes to `Armament`. https://github.com/OpenHV/OpenHV/pull/222

**Comment by dnqbob** [comment]
> We implemented a similar effect without the need for any changes to `Armament`. [OpenHV/OpenHV#222](https://github.com/OpenHV/OpenHV/pull/222)

[OpenHV/OpenHV#222](https://github.com/OpenHV/OpenHV/pull/222) is not ideal to usecase that each barrel firing throws a casing, for example, the mec firing from two sides:
casing-showcase

**Comment by PunkPun** [comment]
Closing as adding proper casing support to third party mods doesn't require any changes to the core engine
